File: algorithms/src/labs/lab_3/F.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.PrintWriter;
import java.util.StringTokenizer;

/**
 * Created by woofi on 20.12.2015.
 */
public class F {
    static class SegmentTree {
        static class Node {
            boolean isFree;
            boolean isLeaf;

            Node() {
                isFree = true;
                isLeaf = false;
            }

            void setLeaf() {
                isLeaf = true;
            }

            void setFree() {
                isFree = true;
            }

            void setBusy() {
                isFree = false;
            }
        }

        int size;
        Node neutral = new Node();
        Node a[];

        SegmentTree(int n) {
            neutral.setLeaf();
            neutral.setBusy();

            size = 1;
            while (size < n) {
                size <<= 1;
            }

            a = new Node[size << 1];

            for (int i = size; i < size + n; i++) {
                a[i] = new Node();
                a[i].setLeaf();
            }
            for (int i = size + n; i < (size << 1); i++) {
                a[i] = neutral;
            }

            for (int i = size; i-- > 1; ) {
                a[i] = new Node();
                if (!(a[i << 1].isFree || a[(i << 1) + 1].isFree))
                {
                    a[i].setBusy();
                }
            }
        }

        int enter(int place) {
            int current = size + place - 1;

            if (a[current].isFree) {
                a[current].setBusy();
                return place;
            }

            int from = size + place - 1;
            int ans = 1 - size;

            while ((current >>= 1) != 0) {
                if ((current << 1) + 1 != from && a[(current << 1) + 1].isFree) {
                    current = (current << 1) + 1;
                    break;
                }
                from = current;
            }
            if (current == 0) {
                current = 1;
            }

            while (!a[current].isLeaf) {
                if (a[current << 1].isFree) {
                    current <<= 1;
                } else {
                    current = (current << 1) + 1;
                }
            }
            a[current].setBusy();
            ans += current;

            while ((current >>= 1) != 0) {
                if (a[current << 1].isFree || a[(current << 1) + 1].isFree) {
                    a[current].setFree();
                } else {
                    a[current].setBusy();
                }
            }

            return ans;
        }

        void exit(int place) {
            int current = size + place - 1;
            a[current].setFree();

            while ((current >>= 1) != 0) {
                a[current].setFree();
            }
        }
    }

    public static void main(String args[]) throws Exception {
        Scanner sc = new Scanner(new File("parking.in"));
        PrintWriter pw = new PrintWriter(new File("parking.out"));

        SegmentTree st = new SegmentTree(sc.nextInt());

        for (int i = sc.nextInt(); i --> 0;) {
            switch (sc.next()) {
                case "enter":
                    pw.println(st.enter(sc.nextInt()));
                    break;
                case "exit":
                    st.exit(sc.nextInt());
                    break;
            }
        }

        sc.close();
        pw.close();
    }

    static class Scanner {
        BufferedReader br;
        StringTokenizer t;

        Scanner(File file) throws Exception {
            br = new BufferedReader(new FileReader(file));
            t = new StringTokenizer("");
        }

        boolean hasNext() throws Exception {
            while (!t.hasMoreTokens()) {
                String line = br.readLine();
                if (line == null)
                    return false;
                t = new StringTokenizer(line);
            }
            return true;
        }

        String next() throws Exception {
            if (hasNext()) {
                return t.nextToken();
            }
            return null;
        }

        int nextInt() throws Exception {
            return Integer.parseInt(next());
        }

        double nextDouble() throws Exception {
            return Double.parseDouble(next());
        }

        void close() throws Exception {
            br.close();
        }
    }
}

File: src/labs/lab_6/F.kt
/**
 * Nariman Safiulin (woofilee)
 * File: F.kt
 * Created on: May 15, 2016
 */

import java.io.BufferedReader
import java.io.File
import java.io.FileReader
import java.io.PrintWriter
import java.util.*

private val PROBLEM_NAME = "negcycle"

private class Scanner(file: File) {
    val br = BufferedReader(FileReader(file))
    var st = StringTokenizer("")

    fun hasNext(): Boolean {
        while (!st.hasMoreTokens()) {
            val line = br.readLine() ?: return false
            st = StringTokenizer(line)
        }
        return true
    }

    fun next(): String? = if (hasNext()) st.nextToken() else null
    fun nextInt(): Int = Integer.parseInt(next())
    fun nextLong(): Long = java.lang.Long.parseLong(next())
    fun nextDouble(): Double = java.lang.Double.parseDouble(next())
    fun close() = br.close()
}

data class DirectedEdge(val from: Int, val to: Int, val weight: Int)

private fun solve(`in`: Scanner, out: PrintWriter) {
    val NON_EXIST_EDGE = 1000000000

    val n = `in`.nextInt()
    val edges = ArrayList<DirectedEdge>()
    val distances = Array(n) { 0 }
    val from = Array(n) { -1 }

    (0..n - 1).forEach { i ->
        (0..n - 1).forEach { j ->
            val w = `in`.nextInt()
            if (w != NON_EXIST_EDGE) {
                edges.add(DirectedEdge(i, j, w))
            }
        }
    }

    var end = -1
    (0..n - 1).forEach { i ->
        end = -1
        edges.forEach { edge ->
            if (distances[edge.to] > distances[edge.from] + edge.weight) {
                distances[edge.to] = Math.max(-NON_EXIST_EDGE, distances[edge.from] + edge.weight)
                from[edge.to] = edge.from
                end = edge.to
            }
        }
    }

    if (end == -1) {
        out.println("NO")
        return
    }

    var start = end
    repeat(n, { start = from[start] })
    val path = ArrayList<Int>()

    var current = start
    while (true) {
        path.add(current)
        if (current == start && path.size > 1)
            break
        current = from[current]
    }

    out.println("YES\n${path.size}")
    (path.size - 1 downTo 0).forEach { i -> out.print("${path[i] + 1} ") }
}

fun main(args: Array<String>) {
    val `in` = Scanner(File(PROBLEM_NAME + ".in"))
    val out = PrintWriter(File(PROBLEM_NAME + ".out"))

    solve(`in`, out)

    `in`.close()
    out.close()
}
    